adapter/service/parameters/model.go

```diff
@@ -0,0 +1,381 @@
+package parameters
+
+import (
+	"fmt"
+)
+
+// ValueType represents a value type.
+type ValueType string
+
+// Constants below represent allowed value types.
+const (
+	ValueTypeInt         ValueType = "int"
+	ValueTypeString      ValueType = "string"
+	ValueTypeBool        ValueType = "bool"
+	ValueTypeIntArray    ValueType = "int_array"
+	ValueTypeStringArray ValueType = "string_array"
+)
+
+func AllowedValueTypes() []ValueType {
+	return []ValueType{
+		ValueTypeInt,
+		ValueTypeString,
+		ValueTypeBool,
+		ValueTypeIntArray,
+		ValueTypeStringArray,
+	}
+}
+
+func IsValueTypeAllowed(t ValueType) bool {
+	for _, allowed := range AllowedValueTypes() {
+		if t == allowed {
+			return true
+		}
+	}
+
+	return false
+}
+
+// WidgetType represents a widget type.
+type WidgetType string
+
+// Constants below represent allowed widget types.
+const (
+	WidgetTypeInput       WidgetType = "input"
+	WidgetTypeSelect      WidgetType = "select"
+	WidgetTypeMultiSelect WidgetType = "multiselect"
+)
+
+var ( //nolint:gofumpt
+	// widgetTypeToValueTypeMapping maps widget types to allowed value types.
+	widgetTypeToValueTypeMapping = map[WidgetType][]ValueType{
+		WidgetTypeInput:       {ValueTypeInt, ValueTypeString, ValueTypeBool},
+		WidgetTypeSelect:      {ValueTypeInt, ValueTypeString},
+		WidgetTypeMultiSelect: {ValueTypeIntArray, ValueTypeStringArray},
+	}
+)
+
+// ParameterSpecification represents a parameter specification that must be provided by the Controller.
+type ParameterSpecification struct {
+	ID           string        `json:"parameter_id"`
+	Name         string        `json:"name"`
+	Description  string        `json:"description"`
+	ValueType    ValueType     `json:"value_type"`
+	WidgetType   WidgetType    `json:"widget_type"`
+	Options      SelectOptions `json:"options,omitempty"`
+	Min          int           `json:"min,omitempty"`
+	Max          int           `json:"max,omitempty"`
+	DefaultValue any           `json:"default_value"`
+	ReadOnly     bool          `json:"read_only"`
+}
+
+// ValidateParameter validates a parameter against the specification.
+func (s ParameterSpecification) ValidateParameter(p Parameter) error {
+	if err := s.validateTypeMatching(p); err != nil {
```

Comment:
THis entire logic of validation is super complicated where it does not have to be. There are simply three validations.

* Does the parameter value type match the spec value type
* If the widget type is Input and value type is int, check min max
* If the widget type is select or multiselect, check if value(s) are contained in options